server/api/jurisdictions.py

```diff
@@ -804,9 +814,110 @@ def get_batch_comparison_audit_discrepancies_by_jurisdiction(
-                "discrepancies": vote_deltas,
+                "reportedVotes": normalize_vals_for_discrepancies(
+                    reported_contest_result
+                ),
+                "auditedVotes": normalize_vals_for_discrepancies(
+                    audited_contest_result
+                ),
+                "discrepancies": normalize_vals_for_discrepancies(vote_deltas),
             }
 
     return discrepancies_by_jurisdiction
+
+
+def get_ballot_comparison_audit_discrepancies_by_jurisdiction(
+    election: Election, round_id: str
+) -> DiscrepanciesByJurisdiction:
+    discrepancies_by_jurisdiction: DiscrepanciesByJurisdiction = defaultdict(
+        lambda: defaultdict(dict)
+    )
+
+    ballots_in_round = (
+        SampledBallot.query.join(SampledBallotDraw)
+        .filter_by(round_id=round_id)
+        .distinct(SampledBallot.id)
+        .with_entities(SampledBallot.id)
+        .subquery()
+    )
+    sampled_ballot_id_to_jurisdiction_id = dict(
+        SampledBallot.query.filter(SampledBallot.id.in_(ballots_in_round))
+        .join(Batch)
+        .with_entities(SampledBallot.id, Batch.jurisdiction_id)
+    )
+    # make a readable identifier - it is the same default format as the ImprintedID
+    # but some sampled ballots have ImprintedIDs of a different format that are
+    # very long and unreadable, so here we ensure they all have the same format
+    # Ex. "Tabulator X, Batch Y, Ballot Position Z, Container 0"
+    sampled_ballot_id_to_readable_identifier = dict(
+        (
+            sampled_ballot_id,
+            f"{tabulator}, {name}, Ballot Position {ballot_position}"
+            + (f", Container {container}" if container is not None else ""),
+        )
+        for sampled_ballot_id, tabulator, name, ballot_position, container in SampledBallot.query.filter(
+            SampledBallot.id.in_(ballots_in_round)
+        )
+        .join(Batch)
+        .with_entities(
+            SampledBallot.id,
+            Batch.tabulator,
+            Batch.name,
+            SampledBallot.ballot_position,
+            Batch.container,
+        )
+    )
+
+    for contest in election.contests:
+        audited_results = sampled_ballot_interpretations_to_cvrs(contest)
+        reported_results = cvrs_for_contest(contest)
+        for ballot_id, audited_result in audited_results.items():
+            audited_cvr = audited_result["cvr"]
+            reported_cvr = reported_results.get(ballot_id)
+            vote_deltas = ballot_vote_deltas(contest, reported_cvr, audited_cvr)
+            if not vote_deltas or isinstance(vote_deltas, str):
+                continue
+
+            reported_votes = {}
+            if reported_cvr and contest.id in reported_cvr:
+                reported_votes = reported_cvr[contest.id]
+
+            audited_votes = {}
+            if audited_cvr and contest.id in audited_cvr:
+                audited_votes = audited_cvr[contest.id]
```

Comment:
I think `reported_cvr` and `audited_cvr` are guaranteed to be non-null at this point due to the checks in `ballot_vote_deltas`, so these could probably be condensed to:

```py
reported_votes = reported_cvr.get(contest.id, {})
audited_votes = audited_cvr.get(contest.id, {})
```

Reply:
I meant they are logically guaranteed to not be None. I know the type system can't figure that out, since the checks to handle the None cases are in a different function. Up to you whether you prefer handling an impossible case to satisfy the type-system or overriding with a type assertion (I usually go for the latter, but that's just my style). Totally fine to leave it as is

Reply:
Oh interesting. Thanks for the clarification. I'll add in the type assertion